### PR TITLE
Adds feature flag to configure `MaxFailedMountAttempts`

### PR DESCRIPTION
**Pull request by meik99**
# Description

Adds a feature flag to configure the value for the `MaxFailedMountAttempts` column and modifies the provisioner so that it is correctly persisted.

## How can this be tested?

* Deploy the operator and cloudnative dynakube
* Copy the `/data/csi.db` from a csi driver pod
* Make sure the default value is used in the dynakube entry of the dynakubes table
* Add the `feature.dynatrace.com/max-csi-mount-attempts` annotation and set it to any value
* If you set it to an invalid value, e.g. a letter or a number < 0, "0" is persisted
* If no value is set, the default value is persisted
* Otherwise the specified value should is persisted

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly
- [x] I have read and understood the [contribution guidelines](https://github.com/Dynatrace/dynatrace-operator/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: src/controllers/csi/metadata/sqlite.go [hunk 1]
```diff
@@ -308,13 +309,18 @@ func (access *SqliteAccess) GetDynakube(dynakubeName string) (*Dynakube, error)
...
+	var maxFailedMountAttempts *int
 
 	err := access.querySimpleStatement(getDynakubeStatement, dynakubeName, &tenantUUID, &latestVersion, &imageDigest, &maxFailedMountAttempts)
 	if err != nil {
 		err = errors.WithMessagef(err, "couldn't get dynakube, name '%s'", dynakubeName)
 	}
-	return NewDynakube(dynakubeName, tenantUUID, latestVersion, imageDigest, maxFailedMountAttempts), err
+
+	if maxFailedMountAttempts == nil {
+		maxFailedMountAttempts = address.Of(-1)
```
Comment (0sewa0) [inline src/controllers/csi/metadata/sqlite.go:320]: shouldn't this be `defaultSqlMaxFailedMountAttempts`?
also if you have `ADD COLUMN MaxFailedMountAttempts INT DEFAULT ' + defaultSqlMaxFailedMountAttempts + ';` wouldn't it always return that default value when reading from the database ?

Comment (0sewa0) [reply src/controllers/csi/metadata/sqlite.go:320]: also, setting the `DEFAULT` in the database feels scary a bit.

what happens if the default changes between releases or something, will the `ALTER` statement be able to handle that ?

File: src/api/v1beta1/properties.go [hunk 2]
```diff
@@ -400,3 +400,11 @@ func getRawImageTag(imageURI string) string {
 	splitURI := strings.Split(imageURI, ":")
 	return splitURI[len(splitURI)-1]
 }
+
+func (dk *DynaKube) MaxFailedCsiMountAttempts() int {
+	if dk.FeatureMaxFailedCsiMountAttempts() != nil {
+		return *dk.FeatureMaxFailedCsiMountAttempts()
+	}
+
+	return -1
```
Comment (0sewa0) [inline src/api/v1beta1/properties.go:409]: wouldn't it be cleaner if we returned the default value here? (and not handled it in the `metadata.NewDynakube()`)

File: src/controllers/csi/metadata/sqlite.go (hunk 1 above)
Comment (meik99) [reply src/controllers/csi/metadata/sqlite.go:320]: The ticket mentioned the column to be nullable.
The default value is only used if the value is not specified on an insert, if null is explicitly inserted the value will be null
If the value is null, it has been explicitly inserted, returning the default value in this case would make the column being nullable useless. I assume the intention is that if null, or -1, is returned, that there is effectively no maximum retry value and it should retry endlessly.
If the default is changed the alter command will handle that accordingly

Comment (meik99) [reply src/controllers/csi/metadata/sqlite.go:320]: after talking about it, it turned out that you were completely correct, I will make the changes accordingly

Comment (meik99) [reply src/controllers/csi/metadata/sqlite.go:320]: done

File: src/api/v1beta1/properties.go (hunk 2 above)
Comment (meik99) [reply src/api/v1beta1/properties.go]: done

File: src/api/v1beta1/feature_flags.go [hunk 3]
```diff
@@ -276,3 +282,19 @@ func (dk *DynaKube) getFeatureFlagRaw(annotation string) string {
...
+	maxCsiMountAttemptsValue := dk.getFeatureFlagRaw(AnnotationFeatureMaxFailedCsiMountAttempts)
+
+	if maxCsiMountAttemptsValue == "" {
+		return defaultMaxFailedCsiMountAttempts
+	}
+
+	maxCsiMountAttempts, err := strconv.Atoi(maxCsiMountAttemptsValue)
+
+	if err != nil || maxCsiMountAttempts < 0 {
+		return 0
```
Comment (0sewa0) [inline src/api/v1beta1/feature_flags.go:296]: shouldn't this be also `defaultMaxFailedCsiMountAttempts`?

File: src/controllers/csi/metadata/metadata_test.go [hunk 4]
```diff
@@ -42,7 +42,7 @@ func TestNewDynakube(t *testing.T) {
 		assert.Equal(t, testUUID, dynakube.TenantUUID)
 		assert.Equal(t, testVersion, dynakube.LatestVersion)
 		assert.Equal(t, testDigest, dynakube.ImageDigest)
-		assert.Equal(t, defaultMaxFailedMountAttempts, dynakube.MaxFailedMountAttempts)
+		assert.Equal(t, -1, dynakube.MaxFailedMountAttempts)
```
Comment (0sewa0) [inline src/controllers/csi/metadata/metadata_test.go:45]: this should never happen

File: src/controllers/csi/metadata/sqlite.go [hunk 5]
```diff
@@ -49,7 +50,7 @@ const (
 
 	dynakubesAlterStatementMaxFailedMountAttempts = `
 	ALTER TABLE dynakubes
-	ADD COLUMN MaxFailedMountAttempts INT DEFAULT ` + defaultSqlMaxFailedMountAttempts + `;`
+	ADD COLUMN MaxFailedMountAttempts INT NOT NULL DEFAULT 0;`  // "Not null"-columns need a default value set
```
Comment (0sewa0) [inline src/controllers/csi/metadata/sqlite.go:53]: spooky to have it at 0, but that should never happen

File: src/controllers/csi/metadata/sqlite.go (hunk 1 above)
Comment (0sewa0) [inline src/controllers/csi/metadata/sqlite.go:320]: `maxFailedMountAttempts` should never be `nil` at this point
-1 is not a valid value

**Review by 0sewa0 — CHANGES_REQUESTED** [review]

File: src/api/v1beta1/feature_flags.go (hunk 3 above)
Comment (meik99) [reply src/api/v1beta1/feature_flags.go:296]: done

File: src/controllers/csi/metadata/metadata_test.go (hunk 4 above)
Comment (meik99) [reply src/controllers/csi/metadata/metadata_test.go:45]: removed the test

File: src/controllers/csi/metadata/sqlite.go (hunk 5 above)
Comment (meik99) [reply src/controllers/csi/metadata/sqlite.go:53]: should i set it to a different value?

File: src/controllers/csi/metadata/sqlite.go (hunk 1 above)
Comment (meik99) [reply src/controllers/csi/metadata/sqlite.go:320]: changed from pointer to value, good catch

File: src/controllers/csi/metadata/sqlite.go (hunk 5 above)
Comment (meik99) [reply src/controllers/csi/metadata/sqlite.go]: set it to 3

**Review by 0sewa0 — DISMISSED** [review]

**Review by 0sewa0 — APPROVED** [review]

**Pull request merged**